### DISCUSSION TITLE
hole punching with dcutr and relay server fails at RemoteInitiatedDirectConnectionUpgrade stage!

### DISCUSSION THREAD

**Discussion by mgharebaghi** [libp2p/rust-libp2p#4212] [Q&A] [answered]
Hello.
I am trying to use dcutr for hole punching by using two clients, one is a listener and the other is a dialer. Everything goes well until the RemoteInitiatedDirectConnectionUpgrade stage, but at this stage the connection fails after a few attempts. Is there a specific reason?
I will be grateful if you guide me. I need it badly

**Comment by thomaseizinger** [comment]
What is your network configuration? Are you behind a punchable NAT? Using TCP or QUIC?

**Comment by mgharebaghi** [comment]
The dialer client is a cloud server with a public address, and the listening client uses a TD LTE modem that I opened NAT through DMZ for the IP used in the modem settings. I use TCP

**Reply by mgharebaghi** [reply]
https://github.com/mgharebaghi82/client-dcurt/blob/main/src/main.rs

**Reply by thomaseizinger** [reply]
> The dialer client is a cloud server with a public address, and the listening client uses a TD LTE modem that I opened NAT through DMZ for the IP used in the modem settings. I use TCP

Mobile ISP usually use carrier-grade NAT that cannot be hole-punched. Can you verify using a traceroute that you are only one hop away from the public internet?

**Reply by mgharebaghi** [reply]
No, I'm more than one hop away from public internet, maybe 5 hops! In such a situation, is it impossible to communicate through hole punching?

**Reply by thomaseizinger** [reply, marked as answer]
If your ISP uses CG-NAT (https://en.m.wikipedia.org/wiki/Carrier-grade_NAT) then it is not possible.

Hole-punching relies on tricking the intermediate NAT device into allowing inbound traffic because of an outbound packet.

If you have more than one NAT device, those would have to use stable port mappings for this to work which is usually not the case. There might be other problems I currently cannot think of.

**Reply by mgharebaghi** [reply]
Thank you for your guidance. Many things became clear to me 🙏
